### PR TITLE
Add some meaningful content to README.md

### PR DESCRIPTION
**Pull request by qmuntal**
This PR updates the README so it contains useful information. Part of it is taken from https://github.com/microsoft/go-crypto-openssl.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,5 +1,63 @@
...
+
+The `openssl` package is designed to be used as a drop-in replacement for the [boring](https://pkg.go.dev/crypto/internal/boring) package in order to facilitate integrating `openssl` inside a forked Go toolchain.
+
+## Disclaimer
+
+A program directly or indirectly using this package in FIPS mode can claim it is using a FIPS-certified cryptographic module (OpenSSL), but it can't claim the program as a whole is FIPS certified without passing the certification process, nor claim it is FIPS compliant without ensuring all crypto APIs and workflows are implemented in a FIPS-compliant manner.
+
+## Background
+
+FIPS 140-2 is a U.S. government computer security standard used to approve cryptographic modules. FIPS compliance may come up when working with U.S. government and other regulated industries.
```
Comment (simo5) [inline README.md:15]: Probably better to mentione FIPS 140-3 at this point.

File: README.md [hunk 2]
```diff
@@ -1,5 +1,63 @@
...
+
+> The status of FIPS 140 for Go itself remains "no plans, basically zero chance".
+
+On the other hand, Google maintains a branch that uses cgo and BoringSSL to implement various crypto primitives: https://github.com/golang/go/blob/dev.boringcrypto/README.boringcrypto.md. As BoringSSL is FIPS 140-2 certified, an application using that branch is more likely to be FIPS 140-2 compliant, yet Google does not provide any liability about the suitability of this code in relation to the FIPS 140-2 standard.
+
+## Features
+
+### Multiple OpenSSL versions supported
+
+The `openssl` package has support for multiple OpenSSL versions, namely 1.0.2, 1.1.0, 1.1.1 and 3.0.x.
```
Comment (simo5) [inline README.md:29]: Should probably say 3.x, given we already have 3.1.x out and 3.2 will come in fall

**Review by simo5 — CHANGES_REQUESTED** [review]
Nits

File: README.md (hunk 2 above)
Comment (qmuntal) [reply README.md:29]: Haven't tested with 3.1.x yet, so I hesitant to add it explicitly.

File: README.md (hunk 1 above)
Comment (qmuntal) [reply README.md:15]: Done

File: README.md [hunk 3]
```diff
@@ -1,5 +1,63 @@
...
+
+All supported OpenSSL versions passes an small set of automatic tests that ensure they can be built and that there are no major regressions.
```
Comment (dagood) [inline README.md:31]:
```suggestion
All supported OpenSSL versions pass a small set of automatic tests that ensure they can be built and that there are no major regressions.
```

File: README.md [hunk 4]
```diff
@@ -1,5 +1,63 @@
...
+These tests do not validate the cryptographic correctness of the `openssl` package.
+
+On top of that, the [Microsoft Go fork](https://github.com/microsoft/go) and the [RedHat Go fork](https://github.com/golang-fips/go) builds and tests a subset of the supported OpenSSL versions when integrated with the Go `crypto` package.
```
Comment (dagood) [inline README.md:34]:
```suggestion
On top of that, the [Microsoft Go fork](https://github.com/microsoft/go) and the [RedHat Go fork](https://github.com/golang-fips/go) build and test a subset of the supported OpenSSL versions when integrated with the Go `crypto` package.
```

File: README.md [hunk 5]
```diff
@@ -1,5 +1,63 @@
...
+
+### Building without OpenSSL headers
+
+The `openssl` package does not use any symbol from the OpenSSL headers. There is no need that have them installed to build an application which imports this library.
+
+CI verifies that all the functions and constants defined in our headers match the ones in the OpenSSL headers for every supported OpenSSL version.
+
+### Portable OpenSSL
+
+The OpenSSL bindings are implemented in such a way that the OpenSSL version used when building a program does not have to match with the OpenSSL version used when running it.
```
Comment (dagood) [inline README.md:45]: We don't really "use" OpenSSL to build, right? (And from there, maybe add another few sentences to clarify how far this feature goes--not only do you not need headers, the library isn't needed either.)

```suggestion
The OpenSSL bindings are implemented in such a way that the OpenSSL version available when building a program does not have to match with the OpenSSL version used when running it.
In fact, OpenSSL doesn't need to be present on the builder.
For example, using the `openssl` package and `go build .` on a Windows host with `GOOS=linux` can produce a program that successfully runs on Linux and uses OpenSSL.
```

File: README.md [hunk 6]
```diff
@@ -1,5 +1,63 @@
...
+
+This feature does not require any additional configuration, but it only works with OpenSSL versions known and supported by the Go toolchain.
```
Comment (dagood) [inline README.md:47]:
```suggestion
This feature does not require any additional configuration, but it only works with OpenSSL versions known and supported by the Go toolchain that integrates the `openssl` package.
```

File: README.md [hunk 7]
```diff
@@ -1,5 +1,63 @@
...
+These tests do not validate the cryptographic correctness of the `openssl` package.
+
+On top of that, the [Microsoft Go fork](https://github.com/microsoft/go) and the [RedHat Go fork](https://github.com/golang-fips/go) builds and tests a subset of the supported OpenSSL versions when integrated with the Go `crypto` package.
+These tests are much more exhaustive and validate a specific OpenSSL version can produce working applications.
+
+### Building without OpenSSL headers
+
+The `openssl` package does not use any symbol from the OpenSSL headers. There is no need that have them installed to build an application which imports this library.
+
+CI verifies that all the functions and constants defined in our headers match the ones in the OpenSSL headers for every supported OpenSSL version.
```
Comment (dagood) [inline README.md:41]:
```suggestion
The CI tests in this repository verify that all the functions and constants defined in our headers match the ones in the OpenSSL headers for every supported OpenSSL version.
```

File: README.md [hunk 8]
```diff
@@ -1,5 +1,65 @@
...
+## Features
+
+### Multiple OpenSSL versions supported
+
+The `openssl` package has support for multiple OpenSSL versions, namely 1.0.2, 1.1.0, 1.1.1 and 3.0.x.
+
+All supported OpenSSL versions pass a small set of automatic tests that ensure they can be built and that there are no major regressions.
+These tests do not validate the cryptographic correctness of the `openssl` package.
+
+On top of that, the [Microsoft Go fork](https://github.com/microsoft/go) and the [RedHat Go fork](https://github.com/golang-fips/go) build and test a subset of the supported OpenSSL versions when integrated with the Go `crypto` package.
```
Comment (dagood) [inline README.md:34]: > RedHat

I believe Red Hat should have a space, or RHEL would refer to the distro. (There's another instance in the `Acknowledgements` section too.)

But I'm not totally sure about calling it Red Hat's fork here, maybe "the golang-fips Go fork". And/or the repo could be updated to mention who's currently maintaining/using it.

Not to say I object to calling it Red Hat's, just want to point this out and IMO it's up to Red Hat folks. 😄

**Review by dagood — APPROVED** [review]
LGTM, one more thing I missed but not totally sure what to change it to.

File: README.md (hunk 8 above)
Comment (qmuntal) [reply README.md]: Good point. I've renamed RedHat to Red Hat and changed the wording to make it clear that golang-fips/go is not Red Hat-specific but it's mainly maintained by them.

**Review by ueno — APPROVED** [review]

**Pull request merged**